nats/src/main/java/io/micronaut/nats/jetstream/KeyValueFactory.java

```diff
@@ -0,0 +1,121 @@
+ * @since 4.0.0
+ */
+@Factory
+public class KeyValueFactory {
+
+    private final BeanContext beanContext;
+
+    @Inject
+    public KeyValueFactory(BeanContext beanContext) {
+        this.beanContext = beanContext;
+    }
+
+
+    /**
+     * @param config The nats connection configuration
+     * @return The keyvalue management
+     * @throws IOException           in case of communication issue
+     * @throws JetStreamApiException the request had an error related to the data
+     */
+    @Singleton
+    @EachBean(NatsConnectionFactoryConfig.class)
+    KeyValueManagement keyValueManagement(NatsConnectionFactoryConfig config) throws IOException, JetStreamApiException {
+        if (config.getJetstream() != null) {
+            KeyValueManagement keyValueManagement = getConnectionByName(config.getName()).keyValueManagement(KeyValueOptions.builder().jetStreamOptions(config.getJetstream().toJetStreamOptions()).build());
+
+            // initialize the given keyvalue configurations
+            for (NatsConnectionFactoryConfig.JetStreamConfiguration.KeyValueConfiguration keyValue : config.getJetstream().getKeyvalue()) {
+                KeyValueConfiguration keyValueConfiguration = keyValue.toKeyValueConfiguration();
+
+                if (keyValueManagement.getBucketNames().contains(keyValueConfiguration.getBucketName())) {
+                    KeyValueStatus status = keyValueManagement.getStatus(keyValueConfiguration.getBucketName());
+                    if (!status.getConfiguration().equals(keyValueConfiguration)) {
+                        keyValueManagement.update(keyValueConfiguration);
+                    }
+                } else {
+                    keyValueManagement.create(keyValueConfiguration);
+                }
+            }
+            return keyValueManagement;
+        }
+        return null;
+    }
+
+    /**
+     * @param injectionPoint injection point
+     * @return The key value bucket
+     * @throws IOException in case of communication issue
+     */
+    @Prototype
+    KeyValue keyvalue(@Nullable InjectionPoint<?> injectionPoint) throws IOException, JetStreamApiException {
+        if (injectionPoint == null) {
+            return null;
+        }
+
+        AnnotationMetadata annotationMetadata = injectionPoint.getAnnotationMetadata();
+
+        if (annotationMetadata.hasAnnotation(KeyValueStore.class)) {
+            String bucketName = annotationMetadata.getAnnotation(KeyValueStore.class)
+                .getRequiredValue(String.class);
+            String connectionName = annotationMetadata.stringValue(NatsConnection.class, "connection")
+                .orElse(NatsConnection.DEFAULT_CONNECTION);
+
+            NatsConnectionFactoryConfig natsConnectionFactoryConfig = beanContext.getBean(NatsConnectionFactoryConfig.class, Qualifiers.byName(connectionName));
+            // Initialize key value management before accessing the key store
+            keyValueManagement(natsConnectionFactoryConfig);
```

Comment:
this is kind of weird, you are calling a method that produces a bean from here?

Reply:
The problem is that the KeyValueManagement needs to be initialized/executed before the keyValue method is called. Otherwise the key/value store might not be created and I'm getting an APIException.

I had similiar problem with the jetstream integration and the `JetstreamIntroductionAdvice`. The `@Requires` annotation did not work in this case. I don't know why, but for me it looks like a race condition

Reply:
Instead of calling the method directly, it is now resolved via BeanContext